R file with a couple of versions of the laplacian interpolator.

richardbeare opened the pull request:
The two functions are regionfill3D and regionfill3DB. They are clones
of the matlab function regionfill, but for 3D.

The driver function is speedimage, which is needs to be given a
distance transform image.

The regionfill3D ends up creating an n x n matrix (sparse), where n is
the number of voxels in the 3D image. regionfill3DB creates an m x m
matrix (sparse), where m is the number of voxels in the mask, after
dilating by 1.

Setting up the matrix structure is messier in regionfill3DB, but the
final matrix will be much smaller. Not yet sure how much practical difference
this will make.

richardbeare committed:
Another version of the laplacian interpolator.

Both versions of the function regionfill3D are clones
of the matlab function regionfill, but for 3D.

The driver function is speedimage, which needs to be given a
distance transform image.

The old regionfill3D ends up creating an n x n matrix (sparse),
where n is the number of voxels in the 3D image.
The new regionfill3D creates an m x m matrix (sparse),
where m is the number of voxels in the mask, after dilating by 1.

Setting up the matrix structure is messier in the new regionfill3D, but the
final matrix will be much smaller. Not yet sure how much practical difference
this will make.

romangrothausmann commented:
Many thanks @richardbeare for this contribution.
I rebased it on the current R branch (with a working DF/DI) and adjusted your commits for direct execution (shebang and u+x) using the script name radius-map.r instead. Is that OK for you?
For being suitable for a GL-CI test, like 
https://github.com/romangrothausmann/SITK-CLIs/blob/830555625d3081aeb6cf25b765a84aa0297e7d5c/.gitlab-ci.yml#L75-L78
it would need some command line args parsing, which is mentioned here: 
https://stackoverflow.com/questions/750786/whats-the-best-way-to-use-r-scripts-on-the-command-line-terminal#comment18803248_969680
Do you have any experience in the use of `commandArgs` function or the `getopt` and `optparse` packages? I've never used such before in R.

romangrothausmann commented:
OK, got a very basic cmd parsing implemented: https://github.com/romangrothausmann/SITK-CLIs/pull/1/commits/7ffcbd5fd491da39fb65cfdbfc622f6f919c95f6
@richardbeare are the adaptions in your commit https://github.com/romangrothausmann/SITK-CLIs/pull/1/commits/e610ac48d6b675f46b0a5916e461538b90561798 correct? Please feel free to adjust as you see fit.

romangrothausmann committed:
test code and extra output:

orig. part of commit f545d7a4acae: "R file with a couple of versions of the laplacian interpolator."

romangrothausmann commented:
Rebased on branch R
Test code removed from orig commit (https://github.com/romangrothausmann/SITK-CLIs/commit/c4bc29495df2c3268eab37d1cafa34a14b2e9809) in acc. to https://github.com/romangrothausmann/SITK-CLIs/pull/1/commits/0f19417599a4121965641c690495d4e16d7f2b11 and added as a separate commit (https://github.com/romangrothausmann/SITK-CLIs/pull/1/commits/8d5e82538493b502454f1815308982409b9975a2) for easing removal when test phase has finished.

romangrothausmann committed:
Revert "test code and extra output:" (until additionally needed test data is available)

This reverts commit ce4ff9cff64a1afa6500d2f4ece80b5d8109ebbf.

romangrothausmann commented:
Rebased again for making use of the GL-CI test now comparing the output with the expected MD5 and providing the result as [`output.mha` artifact](https://gitlab.com/romangrothausmann/SITK-CLIs/-/jobs/202980792/artifacts/browse)

@richardbeare 
What is the meaning of the voxel value in the output? It does not seem to correspond to the local radius of the corr. centerline max. inscribed sphere, as they are mostly > 1e6 while the voxel spacing is 1.
Are voxels values like @ 27 32 33 higher (2.55e7) because a skeleton line would extent towards a bump in that direction?

The modifications of https://github.com/romangrothausmann/SITK-CLIs/pull/1/commits/1abce0110a13bc509cb1d56a6999aaef56015c0e do seem to make a small difference in performance (9s for processing `tests/data/ACN_rs+1.mha` (DimSize = 71 66 45) with https://github.com/romangrothausmann/SITK-CLIs/pull/1/commits/5dde9614ccb3947783f28dfafb7cf9bb2b549a0f and 7s with https://github.com/romangrothausmann/SITK-CLIs/pull/1/commits/1abce0110a13bc509cb1d56a6999aaef56015c0e). The MD5 stays the same despite the use of the new `regionfill3DB` (which is great). However, on the full test data set `tests/data/ACN.mha` (DimSize = 217 202 137) even the new version of `radius-map.r` does not finish within hours.

romangrothausmann commented:
Rebased onto master (R branch now merged into master).

@richardbeare In order to check if I can spot options for optimization concerning the runtime I tried to compare the old `regionfill3D` and the new `regionfill3DB`. To better see the differences between the old `regionfill3D` and the new `regionfill3DB` I removed the old and adjusted the naming of  the new version of `regionfill3D` to reduce the word-diff to a minimum: https://github.com/romangrothausmann/SITK-CLIs/pull/1/commits/d99b0a0931364be2c80d2ca2262f268726f4bb93 
Please check if I made any mistakes there. The result stayed the same.

So far have no idea how to improve the performance, but I guess the hight voxel values originate from
https://github.com/romangrothausmann/SITK-CLIs/blob/d99b0a0931364be2c80d2ca2262f268726f4bb93/radius-map.r#L87-L88
Why multiply with `1e5`?

romangrothausmann commented:
Another rebase to get some timing of the executed command during testing, e.g. compare:
https://gitlab.com/romangrothausmann/SITK-CLIs/-/jobs/203080939
with
https://gitlab.com/romangrothausmann/SITK-CLIs/-/jobs/203081357

romangrothausmann commented:
Tests show that the old version takes around 19s and the new one around 15s, so going for the new one.
